spark/agent/src/main/scala/za/co/absa/spline/harvester/builder/write/WriteCommandExtractor.scala

```diff
@@ -19,73 +19,84 @@ package za.co.absa.spline.harvester.builder.write
-          val jdbcConnectionString = cmd.options("url")
-          val tableName = cmd.options("dbtable")
-          WriteCommand(cmd.nodeName, SourceIdentifier.forJDBC(jdbcConnectionString, tableName), cmd.mode, cmd.query)
-        case _ =>
-          val maybeFormat = maybeSourceType.map {
-            case dsr: DataSourceRegister => dsr.shortName
-            case o => o.toString
-          }
-          val opts = cmd.options
-          val uri = opts.get("path").map(pathQualifier.qualify)
-            .orElse(opts.get("topic").filter(_ => opts.contains("kafka.bootstrap.servers")).map(SourceUri.forKafka))
-            .getOrElse(sys.error(s"Cannot extract source URI from the options: ${opts.keySet mkString ","}"))
-          WriteCommand(cmd.nodeName, SourceIdentifier(maybeFormat, uri), cmd.mode, cmd.query, opts)
-      }
-
-    case cmd: InsertIntoHadoopFsRelationCommand =>
-      val path = cmd.outputPath.toString
-      val format = cmd.fileFormat.toString
-      val qPath = pathQualifier.qualify(path)
-      WriteCommand(cmd.nodeName, SourceIdentifier(Some(format), qPath), cmd.mode, cmd.query, cmd.options)
-
-    case `_: InsertIntoDataSourceDirCommand`(cmd) =>
-      asDirWriteCommand(cmd.nodeName, cmd.storage, cmd.provider, cmd.overwrite, cmd.query)
-
-    case `_: InsertIntoHiveDirCommand`(cmd) =>
-      asDirWriteCommand(cmd.nodeName, cmd.storage, "hive", cmd.overwrite, cmd.query)
-
-    case `_: InsertIntoHiveTable`(cmd) =>
-      val mode = if (cmd.overwrite) Overwrite else Append
-      asTableWriteCommand(cmd.nodeName, cmd.table, mode, cmd.query)
-
-    case `_: CreateHiveTableAsSelectCommand`(cmd) =>
-      val sourceId = SourceIdentifier.forTable(cmd.tableDesc)(pathQualifier, session)
-      WriteCommand(cmd.nodeName, sourceId, cmd.mode, cmd.query)
-
-    case cmd: CreateDataSourceTableAsSelectCommand =>
-      asTableWriteCommand(cmd.nodeName, cmd.table, cmd.mode, cmd.query)
-
-    case dtc: DropTableCommand =>
-      val uri = SourceUri.forTable(dtc.tableName)(session)
-      val sourceId = SourceIdentifier(None, pathQualifier.qualify(uri))
-      WriteCommand(dtc.nodeName, sourceId, Overwrite, dtc)
-
-    case ctc: CreateTableCommand =>
-      val sourceId = SourceIdentifier.forTable(ctc.table)(pathQualifier, session)
-      WriteCommand(ctc.nodeName, sourceId, Overwrite, ctc)
+class WriteCommandExtractor(pathQualifier: PathQualifier, session: SparkSession, splineMode: SplineMode) {
+
+  private val logger = LoggerFactory.getLogger(getClass.getSimpleName)
+
+  def asWriteCommand(operation: LogicalPlan): Option[WriteCommand] = {
+    val maybeWriteCommand = condOpt(operation) {
+      case cmd: SaveIntoDataSourceCommand =>
+        val maybeSourceType = DataSourceTypeExtractor.unapply(cmd)
+        maybeSourceType match {
+          case Some(sourceType) if sourceType == "jdbc" || sourceType.isInstanceOf[JdbcRelationProvider] =>
```

Comment:
These are quite big, do you think extracting some of the cases would make the code more readable?

Reply:
I'd say it's alright. Considering the fact that the match is far from being complete and that there is (still) a large unknown area for us in terms of finding a right implementation of those cases (existing ones and the ones that will be added in next versions), I'd prefer to keep them close together to have better view on all at once, and to facilitate the ongoing refactoring during further development. We'll probably decompose this class later when we have better understanding of how those pieces should interact, what can/should be reused and so forth. For now it's in the state of ongoing development.